core/commands/filestore.go

```diff
@@ -37,58 +48,89 @@ The output is:
 <hash> <size> <path> <offset>
 `,
 	},
-	Arguments: []cmds.Argument{
-		cmds.StringArg("obj", false, true, "Cid of objects to list."),
+	Arguments: []cmdsutil.Argument{
+		cmdsutil.StringArg("obj", false, true, "Cid of objects to list."),
 	},
-	Run: func(req cmds.Request, res cmds.Response) {
-		_, fs, err := getFilestore(req)
+	Run: func(req cmds.Request, re cmds.ResponseEmitter) {
+		_, fs, err := getFilestore(req.InvocContext())
 		if err != nil {
-			res.SetError(err, cmds.ErrNormal)
+			re.SetError(err, cmdsutil.ErrNormal)
 			return
 		}
 		args := req.Arguments()
 		if len(args) > 0 {
 			out := perKeyActionToChan(args, func(c *cid.Cid) *filestore.ListRes {
 				return filestore.List(fs, c)
 			}, req.Context())
-			res.SetOutput(out)
+
+			for v := range out {
+				re.Emit(v)
+			}
 		} else {
 			next, err := filestore.ListAll(fs)
 			if err != nil {
-				res.SetError(err, cmds.ErrNormal)
+				re.SetError(err, cmdsutil.ErrNormal)
 				return
 			}
+
 			out := listResToChan(next, req.Context())
-			res.SetOutput(out)
+			for v := range out {
+				log.Debugf("%T", v)
+				re.Emit(v)
+			}
 		}
 	},
-	Marshalers: cmds.MarshalerMap{
-		cmds.Text: func(res cmds.Response) (io.Reader, error) {
-			outChan, ok := res.Output().(<-chan interface{})
-			if !ok {
-				return nil, u.ErrCast()
-			}
-			errors := false
-			for r0 := range outChan {
-				r := r0.(*filestore.ListRes)
-				if r.ErrorMsg != "" {
-					errors = true
-					fmt.Fprintf(res.Stderr(), "%s\n", r.ErrorMsg)
+	PostRun: map[cmds.EncodingType]func(cmds.Request, cmds.ResponseEmitter) cmds.ResponseEmitter{
+		cmds.CLI: func(req cmds.Request, re cmds.ResponseEmitter) cmds.ResponseEmitter {
+			re_, res := cmds.NewChanResponsePair(req)
+
+			go func() {
+				defer re.Close()
+
+				var (
+					err    error
+					errors bool
+				)
+
+				for err == nil {
```

Comment:
I'm not sure how a switch would help here but I made a change that removes the condition from the loop head